### PR TITLE
[DNM] tikv-ctl cleanup a range

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**

### CODE REVIEW

File: src/bin/tikv-ctl.rs [hunk 1]
```diff
@@ -673,6 +690,15 @@ impl DebugExecutor for Debugger {
         eprintln!("only support remote mode");
         process::exit(-1);
     }
+
+    fn do_cleanup(&self, from: Vec<u8>, to: Vec<u8>) {
+        let db = self.get_db_from_type(DBType::KV).unwrap();
+        if let Err(e) = delete_all_in_range(db, &from, &to, false) {
```
Comment (siddontang) [inline src/bin/tikv-ctl.rs]: this may block the gRPC thread.

File: src/bin/tikv-ctl.rs [hunk 2]
```diff
@@ -409,6 +411,15 @@ trait DebugExecutor {
 
     fn check_local_mode(&self);
 
+    fn cleanup(&self, from: Option<&str>, to: Option<&str>) {
```
Comment (hicqu) [inline src/bin/tikv-ctl.rs]: We don't need `Option` if we set 2 arguments are `required`.

File: src/bin/tikv-ctl.rs [hunk 3]
```diff
@@ -1069,6 +1095,24 @@ fn main() {
...
                  .help("meta file path"))
+        )
+        .subcommand(
+            SubCommand::with_name("cleanup")
+                .about("clean up a range for all column families of kv db")
+                .arg(
+                    Arg::with_name("from")
+                        .short("f")
+                        .long("from")
+                        .takes_value(true)
```
Comment (hicqu) [inline src/bin/tikv-ctl.rs]: you can add `.required(true)` here to `unwrap` it safely.

File: src/bin/tikv-ctl.rs [hunk 4]
```diff
@@ -1192,6 +1236,10 @@ fn main() {
     } else if let Some(matches) = matches.subcommand_matches("metrics") {
         let tags = Vec::from_iter(matches.values_of("tag").unwrap());
         debug_executor.dump_metrics(tags)
+    } else if let Some(matches) = matches.subcommand_matches("cleanup") {
+        let from = matches.value_of("from");
```
Comment (hicqu) [inline src/bin/tikv-ctl.rs]: I think `unwrap` here is better than in `cleanup`.

**Comment by zhangjinpeng1987** [comment]
I don't know if it is necessary and safe to cleanup a range for a running TiKV.

File: src/bin/tikv-ctl.rs [hunk 5]
```diff
@@ -409,6 +411,12 @@ trait DebugExecutor {
 
     fn check_local_mode(&self);
 
+    fn cleanup(&self, from: &str, to: &str) {
```
Comment (MyonKeminta) [inline src/bin/tikv-ctl.rs]: Isn't it better to name the parameters `start_key` and `end_key` ?

**Comment by MyonKeminta** [comment]
I briefly tested the correctness this feature.
I generated some tables by using `importer` provided in tidb-tools, and get their `table_id` from TiDB's report port. Then I delete all data keys of a table by using the new "cleanup" feature, and check if the table's content is gone and other tables are not affected. To check whether a table's content is changed or not, I queried some counts and sums as a simple summary and check whether these query results changes.

## The test database
DDL:
```sql
create table if not exists t(a int primary key, b double, c varchar(10), d date, e time, f timestamp, g date, h datetime);
create index u_b on t(b);
```

Tables and table ids:
```
Table:"test.t4",     TableID:40 (8000000000000028)
Table:"test.t5",     TableID:43 (800000000000002b)
Table:"test.t6",     TableID:46 (800000000000002e)
Table:"test.t7",     TableID:49 (8000000000000031)
Table:"test.t8",     TableID:52 (8000000000000034)
Table:"test.t9",     TableID:55 (8000000000000037)
```

Before testing:
```sql
-- This query is used to produce a simple summary of the table, to check if the table is modified after running cleanup
select count(*), avg(a), sum(b), avg(g) from t;
```
|table| count(*)   | avg(a)            |  sum(b)             | avg(g)          |
|----|------------|----------------|-------------------|------------------|
|t4:  | 18615000 | 9307499.5000 | 2249134023554 | 20180664.4983 |
|t5:  |    34000 | 72490.9028 | -201794161498 | 20180661.9880 |
|t6:  | 986069000 | 493034560.6651 | 39227992596350 | 20180664.5189 |
|t7:  | 52707000 | 26353499.5000 | 10782881608355 | 20180664.4838 |
|t8:  |  3325000 | 1662880.0304 | -1787078957472 | 20180664.4561 |
|t9:  | 43264000 | 21631999.5000 | 6812665116098 | 20180664.5455 |

## Deleting content of table t5

The key range of t5 is [0x74800000000000002b, 0x74800000000000002c)

For all TiKV instances of my test cluster, do clean up:
```
./tikv-ctl --db /data1/misono/kv1/deploy/data/db/ cleanup --from 74800000000000002b --to 74800000000000002c
./tikv-ctl --db /data1/misono/kv2/deploy/data/db/ cleanup --from 74800000000000002b --to 74800000000000002c
./tikv-ctl --db /data2/misono/kv3/deploy/data/db/ cleanup --from 74800000000000002b --to 74800000000000002c
./tikv-ctl --db /data2/misono/kv4/deploy/data/db/ cleanup --from 74800000000000002b --to 74800000000000002c
```

After cleaning up:

|table| count(*)   | avg(a)            |  sum(b)             | avg(g)          |
|----|------------|----------------|-------------------|------------------|
|t4:  | 18615000 | 9307499.5000 | 2249134023554 | 20180664.4983 |
|t5:  |               0|                      - |                          - |                        - |
|t6:  | 986069000 | 493034560.6651 | 39227992596350 | 20180664.5189 |

All data in t5 disappeared. t4 and t6 was not affected.

## Deleting content of table t8
The key range of t8 is [0x748000000000000034, 0x748000000000000035)

```
./tikv-ctl --db /data1/misono/kv1/deploy/data/db/ cleanup --from 748000000000000034 --to 748000000000000035
./tikv-ctl --db /data1/misono/kv2/deploy/data/db/ cleanup --from 748000000000000034 --to 748000000000000035
./tikv-ctl --db /data2/misono/kv3/deploy/data/db/ cleanup --from 748000000000000034 --to 748000000000000035
./tikv-ctl --db /data2/misono/kv4/deploy/data/db/ cleanup --from 748000000000000034 --to 748000000000000035
```

After cleaning up:

|table| count(*)   | avg(a)            |  sum(b)             | avg(g)          |
|----|------------|----------------|-------------------|------------------|
|t7:  | 52707000 | 26353499.5000 | 10782881608355 | 20180664.4838 |
|t8:  |               0 |                      -  |                          -  |                       - |
|t9:  | 43264000 | 21631999.5000 | 6812665116098 | 20180664.5455 |

All data in t8 disappeared. t7 and t9 was not affected.

## Conclusion
The behavior of the feature is the same as expected.
I didn't find any problem on correctness.

**Comment by siddontang** [comment]
Great Job @MyonKeminta

**Comment by siddontang** [comment]
@zhangjinpeng1987 we should support remote mode, we can't let the user stop their service and do the cleanup.

**Pull request closed without merging**